### ISSUE TITLE
angr/claripy fail to catch zero division error

### ISSUE THREAD

**Issue by mirokuxy** [angr/angr#773]
angr/claripy failed to catch zero division error (in c code and in claripy)

the c code I was executing:
```
    // argv[0] set as symbolic BV16
    char x = argv[0][0];
    y = x < 'e' ? 0 : 1;
    y = 1 / y;
```

the angr code:
```
proj = angr.Project('binary', auto_load_libs=True)
x = state.solver.BVS('x', 16)
state = proj.factory.entry_state(args=[x])
simgr = proj.factory.simulation_manager(state)
simgr.run()
```

the complete error message:
```
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-191-d7a0b3e79488> in <module>()
----> 1 simgr.run()

/fuzzer/venv/local/lib/python2.7/site-packages/angr/manager.pyc in run(self, stash, n, step_func)
    935
    936         until_func = lambda pg: any(h(pg) for h in self._hooks_complete)
--> 937         return self.step(n=n, step_func=step_func, until=until_func, stash=stash)
    938
    939

/fuzzer/venv/local/lib/python2.7/site-packages/angr/manager.pyc in step(self, n, selector_func, step_func, stash, successor_func, until, **kwargs)
    598             l.debug("Round %d: stepping %s", i, pg)
    599
--> 600             pg = pg._one_step(stash=stash, selector_func=selector_func, successor_func=successor_func, **kwargs)
    601             if step_func is not None:
    602                 pg = step_func(pg)

/fuzzer/venv/local/lib/python2.7/site-packages/angr/manager.pyc in _one_step(self, stash, selector_func, successor_func, **kwargs)
    375         # each tick produces a dict of stashes. use _record_step_results to dump them into the result pool.
    376         for a in to_tick:
--> 377             result_stashes = self._one_state_step(a, successor_func=successor_func, **kwargs)
    378             self._record_step_results(new_stashes, new_active, result_stashes)
    379

/fuzzer/venv/local/lib/python2.7/site-packages/angr/manager.pyc in _one_state_step(self, a, successor_func, resilience, **kwargs)
    253                     ss = successor_func(a)
    254                 else:
--> 255                     ss = self._project.factory.successors(a, **kwargs)
    256
    257                 new_stashes = self._make_stashes_dict(

/fuzzer/venv/local/lib/python2.7/site-packages/angr/factory.pyc in successors(self, state, addr, jumpkind, inline, default_engine, engines, **kwargs)
     75         for engine in engines:
     76             if engine.check(state, inline=inline, **kwargs):
---> 77                 r = engine.process(state, inline=inline,**kwargs)
     78                 if r.processed:
     79                     break

/fuzzer/venv/local/lib/python2.7/site-packages/angr/engines/hook.pyc in process(self, state, procedure, force_addr, **kwargs)
     47
     48         l.debug("Running %s (originally at %#x)", repr(procedure), addr)
---> 49         return self.project.factory.procedure_engine.process(state, procedure, force_addr=force_addr, **kwargs)
     50
     51     #

/fuzzer/venv/local/lib/python2.7/site-packages/angr/engines/procedure.pyc in process(self, state, procedure, ret_to, inline, force_addr, **kwargs)
     32                 ret_to=ret_to,
     33                 inline=inline,
---> 34                 force_addr=force_addr)
     35
     36     def _check(self, state, *args, **kwargs):

/fuzzer/venv/local/lib/python2.7/site-packages/angr/engines/engine.pyc in process(***failed resolving arguments***)
     52         successors = new_state._inspect_getattr('sim_successors', successors)
     53         try:
---> 54             self._process(new_state, successors, *args, **kwargs)
     55         except SimException:
     56             if o.EXCEPTION_HANDLING not in old_state.options:

/fuzzer/venv/local/lib/python2.7/site-packages/angr/engines/procedure.pyc in _process(self, state, successors, procedure, ret_to)
     66
     67         # do it
---> 68         inst = procedure.execute(state, successors, ret_to=ret_to)
     69         successors.artifacts['procedure'] = inst
     70

/fuzzer/venv/local/lib/python2.7/site-packages/angr/sim_procedure.pyc in execute(self, state, successors, arguments, ret_to)
    155
    156             # run it
--> 157             r = getattr(inst, inst.run_func)(*sim_args, **inst.kwargs)
    158
    159         if inst.returns and (not inst.successors or len(inst.successors.successors) == 0):

/fuzzer/venv/local/lib/python2.7/site-packages/angr/procedures/libc/printf.pyc in run(self)
     14         # The format str is at index 0
     15         fmt_str = self._parse(0)
---> 16         out_str = fmt_str.replace(1, self.arg)
     17
     18         self.state.posix.write(1, out_str, out_str.size() / 8)

/fuzzer/venv/local/lib/python2.7/site-packages/angr/procedures/stubs/format_parser.pyc in replace(self, startpos, args)
     68                 else:
     69                     i_val = args(argpos)
---> 70                     c_val = int(self.parser.state.se.eval(i_val))
     71                     c_val &= (1 << (fmt_spec.size * 8)) - 1
     72                     if fmt_spec.signed and (c_val & (1 << ((fmt_spec.size * 8) - 1))):

/fuzzer/venv/local/lib/python2.7/site-packages/angr/state_plugins/solver.pyc in eval(self, e, **kwargs)
    523         """
    524         # eval_upto already throws the UnsatError, no reason for us to worry about it
--> 525         return self.eval_upto(e, 1, **kwargs)[0]
    526
    527     def eval_one(self, e, **kwargs):

/fuzzer/venv/local/lib/python2.7/site-packages/angr/state_plugins/solver.pyc in eval_upto(self, e, n, cast_to, **kwargs)
    506             return [self._cast_to(e, concrete_val, cast_to)]
    507
--> 508         cast_vals = [self._cast_to(e, v, cast_to) for v in self._eval(e, n, **kwargs)]
    509         if len(cast_vals) == 0:
    510             raise SimUnsatError('Not satisfiable: %s, expected up to %d solutions' % (e.shallow_repr(), n))

/fuzzer/venv/local/lib/python2.7/site-packages/angr/state_plugins/solver.pyc in concrete_shortcut_tuple(self, *args, **kwargs)
    155         v = _concrete_value(args[0])
    156         if v is None:
--> 157             return f(self, *args, **kwargs)
    158         else:
    159             return ( v, )

/fuzzer/venv/local/lib/python2.7/site-packages/angr/state_plugins/sim_action_object.pyc in ast_stripper(*args, **kwargs)
     51         new_args = _raw_ast(args)
     52         new_kwargs = _raw_ast(kwargs)
---> 53         return f(*new_args, **new_kwargs)
     54     return ast_stripper
     55

/fuzzer/venv/local/lib/python2.7/site-packages/angr/state_plugins/solver.pyc in wrapped_f(*args, **kwargs)
     86     def wrapped_f(*args, **kwargs):
     87         try:
---> 88             return f(*args, **kwargs)
     89         except claripy.UnsatError:
     90             e_type, value, traceback = sys.exc_info()

/fuzzer/venv/local/lib/python2.7/site-packages/angr/state_plugins/solver.pyc in _eval(self, e, n, extra_constraints, exact)
    367         :rtype: tuple
    368         """
--> 369         return self._solver.eval(e, n, extra_constraints=self._adjust_constraint_list(extra_constraints), exact=exact)
    370
    371     @concrete_path_scalar

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/concrete_handler_mixin.pyc in eval(self, e, n, **kwargs)
      5             return (c,)
      6         else:
----> 7             return super(ConcreteHandlerMixin, self).eval(e, n, **kwargs)
      8
      9     def batch_eval(self, exprs, n, **kwargs): #pylint:disable=unused-argument

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/constraint_filter_mixin.pyc in eval(self, e, n, extra_constraints, **kwargs)
     38     def eval(self, e, n, extra_constraints=(), **kwargs):
     39         ec = self._constraint_filter(extra_constraints)
---> 40         return super(ConstraintFilterMixin, self).eval(e, n, extra_constraints=ec, **kwargs)
     41
     42     def batch_eval(self, exprs, n, extra_constraints=(), **kwargs):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/sat_cache_mixin.pyc in eval(self, e, n, extra_constraints, **kwargs)
     54             r = super(SatCacheMixin, self).eval(
     55                 e, n,
---> 56                 extra_constraints=extra_constraints, **kwargs
     57             )
     58             self._cached_satness = True

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/simplify_helper_mixin.pyc in eval(self, e, n, *args, **kwargs)
     11         if n > 1:
     12             self.simplify()
---> 13         return super(SimplifyHelperMixin, self).eval(e, n, *args, **kwargs)
     14
     15     def batch_eval(self, e, n, *args, **kwargs):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/constraint_expansion_mixin.pyc in eval(self, e, n, extra_constraints, exact, **kwargs)
     10             extra_constraints=extra_constraints,
     11             exact=exact,
---> 12             **kwargs
     13         )
     14

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontends/composite_frontend.pyc in eval(self, e, n, extra_constraints, exact)
    282
    283         ms = self._merged_solver_for(e=e, lst=extra_constraints)
--> 284         r = ms.eval(e, n, extra_constraints=extra_constraints, exact=exact)
    285         self._reabsorb_solver(ms)
    286         return r

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/sat_cache_mixin.pyc in eval(self, e, n, extra_constraints, **kwargs)
     54             r = super(SatCacheMixin, self).eval(
     55                 e, n,
---> 56                 extra_constraints=extra_constraints, **kwargs
     57             )
     58             self._cached_satness = True

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/model_cache_mixin.pyc in eval(self, e, n, **kwargs)
    265
    266     def eval(self, e, n, **kwargs):
--> 267         return tuple( r[0] for r in ModelCacheMixin.batch_eval(self, [e], n=n, **kwargs) )
    268
    269     def min(self, e, extra_constraints=(), **kwargs):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/model_cache_mixin.pyc in batch_eval(self, asts, n, extra_constraints, **kwargs)
    236
    237     def batch_eval(self, asts, n, extra_constraints=(), **kwargs):
--> 238         results = self._get_batch_solutions(asts, n=n, extra_constraints=extra_constraints)
    239
    240         if len(results) == n or (len(asts) == 1 and asts[0].cache_key in self._eval_exhausted):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/model_cache_mixin.pyc in _get_batch_solutions(self, asts, n, extra_constraints)
    213
    214         for m in self._get_models(extra_constraints):
--> 215             results.add(m.eval_list(asts))
    216             if len(results) == n:
    217                 break

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/model_cache_mixin.pyc in eval_list(self, asts)
     69
     70     def eval_list(self, asts):
---> 71         return tuple(self.eval_ast(c) for c in asts)
     72
     73 class ModelCacheMixin(object):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/model_cache_mixin.pyc in <genexpr>((c,))
     69
     70     def eval_list(self, asts):
---> 71         return tuple(self.eval_ast(c) for c in asts)
     72
     73 class ModelCacheMixin(object):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/frontend_mixins/model_cache_mixin.pyc in eval_ast(self, ast)
     55         # If there was no last value, it was not constrained, so we can use
     56         # anything.
---> 57         new_ast = ast._replace(self.replacements, leaf_operation=self._leaf_op)
     58         return backends.concrete.eval(new_ast, 1)[0]
     59

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/ast/base.pyc in _replace(self, replacements, variable_set, leaf_operation)
    614                 for a in self.args:
    615                     if isinstance(a, Base):
--> 616                         new_a = a._replace(replacements=replacements, variable_set=variable_set, leaf_operation=leaf_operation)
    617                         replaced |= new_a is not a
    618                     else:

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/ast/base.pyc in _replace(self, replacements, variable_set, leaf_operation)
    614                 for a in self.args:
    615                     if isinstance(a, Base):
--> 616                         new_a = a._replace(replacements=replacements, variable_set=variable_set, leaf_operation=leaf_operation)
    617                         replaced |= new_a is not a
    618                     else:

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/ast/base.pyc in _replace(self, replacements, variable_set, leaf_operation)
    622
    623                 if replaced:
--> 624                     r = self.make_like(self.op, tuple(new_args))
    625                     replacements[hash_key] = r
    626                 else:

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/ast/bits.pyc in make_like(self, *args, **kwargs)
     18     def make_like(self, *args, **kwargs):
     19         if 'length' not in kwargs: kwargs['length'] = self.length
---> 20         return Base.make_like(self, *args, **kwargs)
     21
     22     def size(self):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/ast/base.pyc in make_like(self, *args, **kwargs)
    304         if 'uninitialized' not in kwargs: kwargs['uninitialized'] = self._uninitialized
    305         if 'symbolic' not in kwargs and self.op in all_operations: kwargs['symbolic'] = self.symbolic
--> 306         return type(self)(*args, **kwargs)
    307
    308     def _rename(self, new_name):

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/ast/base.pyc in __new__(cls, op, args, **kwargs)
    132             for eb in eager_backends:
    133                 try:
--> 134                     r = operations._handle_annotations(eb._abstract(eb.call(op, args)), args)
    135                     if r is not None:
    136                         return r

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/backends/__init__.pyc in call(self, op, args)
    197         if op in self._op_raw:
    198             # the raw ops don't get the model, cause, for example, Z3 stuff can't take it
--> 199             obj = self._op_raw[op](*converted)
    200         elif not op.startswith("__"):
    201             l.debug("backend has no operation %s", op)

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/bv.pyc in normalize_helper(self, o)
     38         if not isinstance(self, BVV) or not isinstance(o, BVV):
     39             return NotImplemented
---> 40         return f(self, o)
     41
     42     return normalize_helper

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/bv.pyc in compare_guard(self, o)
     13         if self.bits != o.bits:
     14             raise ClaripyTypeError("bitvectors are differently-sized (%d and %d)" % (self.bits, o.bits))
---> 15         return f(self, o)
     16
     17     return compare_guard

/fuzzer/venv/local/lib/python2.7/site-packages/claripy/bv.pyc in SDiv(self, o)
    405     a = self.signed
    406     b = o.signed
--> 407     val = a//b if a*b>0 else (a+(-a%b))//b
    408     return BVV(val, self.bits)
    409

ZeroDivisionError: long division or modulo by zero
```

**Comment by mirokuxy** [comment]
@rhelmot

**Comment by ltfish** [comment]
Thanks for reporting! Fixed with claripy commit [8ea323c](https://github.com/angr/claripy/commit/8ea323cb3c131f20da56821e561fe724b7feeeae).
